File: edumark-admin/src/api/marking.ts
import { request } from '@/utils/request'
import type { Id, PageResult } from './types'

// 阅卷任务相关接口
export interface MarkingTaskQuery {
  pageNum?: number
  pageSize?: number
  examId?: Id
  examSubjectId?: Id
  status?: number
}

export interface MarkingTaskVO {
  id: Id
  examId: Id
  examName: string
  examSubjectId: Id
  subjectName: string
  questionId: Id
  questionNo: string
  name: string
  taskType: number
  totalCount: number
  completedCount: number
  pendingCount: number
  enableDoubleMarking: number
  doubleMarkingThreshold: number
  status: number
  statusName: string
  accessCode?: string
  secondAccessCode?: string
  accessCodeExpireTime?: string
  startTime: string
  endTime: string
  remark: string
  createTime: string
  assigns?: MarkingTaskAssignVO[]
}

export interface MarkingTaskAssignVO {
  id: Id
  taskId: Id
  taskName: string
  teacherId: Id
  teacherName: string
  assignCount: number
  completedCount: number
  markingRole: number
  markingRoleName: string
  status: number
  statusName: string
  examName: string
  subjectName: string
  questionNo: string
  taskStatus: number
}

export interface TeacherAssign {
  teacherId: Id
  assignCount?: number
  markingRole: number
}

export interface MarkingTaskAssignDTO {
  taskId: Id
  assigns: TeacherAssign[]
}

// 阅卷记录相关接口
export interface MarkingRecordVO {
  id: number
  taskId: number
  answerSheetId: number
  questionId: number
  questionNo: string
  studentId: number
  studentName: string
  studentNumber: string
  teacherId: number
  teacherName: string
  markingRole: number
  markingRoleName: string
  score: number
  fullScore: number
  comment: string
  markingTime: string
  status: number
  statusName: string
  answerImages: string[]
  answerImageUrl: string
  originalImageUrl?: string
  createTime: string
}

export interface MarkingSubmitDTO {
  recordId: number
  score: number
  comment?: string
}

// 仲裁记录相关接口
export interface MarkingArbitrationVO {
  id: number
  taskId: number
  answerSheetId: number
  questionId: number
  questionNo: string
  studentId: number
  studentName: string
  firstScore: number
  firstTeacherId: number
  firstTeacherName: string
  secondScore: number
  secondTeacherId: number
  secondTeacherName: string
  scoreDiff: number
  arbitrationTeacherId: number
  arbitrationTeacherName: string
  arbitrationScore: number
  arbitrationTime: string
  arbitrationComment: string
  status: number
  statusName: string
  answerImages: string[]
  answerImageUrl: string
  originalImageUrl?: string
  createTime: string
}

export interface ArbitrationSubmitDTO {
  arbitrationId: number
  score: number
  comment?: string
}

// 阅卷任务管理 API
export function pageMarkingTasks(params: MarkingTaskQuery) {
  return request.get<PageResult<MarkingTaskVO>>('/marking/task/page', { params })
}

export function getMarkingTaskDetail(id: Id) {
  return request.get<MarkingTaskVO>(`/marking/task/${id}`)
}

export function generateMarkingTasks(examSubjectId: Id) {
  return request.post<void>(`/marking/task/generate/${examSubjectId}`)
}

export function deleteMarkingTask(id: Id) {
  return request.delete<void>(`/marking/task/${id}`)
}

export function assignMarkingTask(data: MarkingTaskAssignDTO) {
  return request.post<void>('/marking/task/assign', data)
}

export function startMarkingTask(id: Id) {
  return request.post<void>(`/marking/task/start/${id}`)
}

export function completeMarkingTask(id: Id) {
  return request.post<void>(`/marking/task/complete/${id}`)
}

export function listMarkingTasksByExamSubject(examSubjectId: Id) {
  return request.get<MarkingTaskVO[]>(`/marking/task/list/${examSubjectId}`)
}

// 阅卷工作台 API
export function getMyAssigns() {
  return request.get<MarkingTaskAssignVO[]>('/marking/my-assigns')
}

export function pageMarkingRecords(params: { taskId: number; status?: number; pageNum?: number; pageSize?: number }) {
  return request.get<PageResult<MarkingRecordVO>>('/marking/records', { params })
}

export function getNextPendingRecord(taskId: Id) {
  return request.get<MarkingRecordVO>('/marking/next-pending', { params: { taskId } })
}

export function getMarkingRecordDetail(recordId: number) {
  return request.get<MarkingRecordVO>(`/marking/record/${recordId}`)
}

export function submitMarkingScore(data: MarkingSubmitDTO) {
  return request.post<void>('/marking/submit', data)
}

// 仲裁 API
export function pageArbitrations(params: { taskId: number; status?: number; pageNum?: number; pageSize?: number }) {
  return request.get<PageResult<MarkingArbitrationVO>>('/marking/arbitrations', { params })
}

export function getNextArbitration(taskId: Id) {
  return request.get<MarkingArbitrationVO>('/marking/next-arbitration', { params: { taskId } })
}

export function getArbitrationDetail(arbitrationId: number) {
  return request.get<MarkingArbitrationVO>(`/marking/arbitration/${arbitrationId}`)
}

export function submitArbitration(data: ArbitrationSubmitDTO) {
  return request.post<void>('/marking/arbitration/submit', data)
}

export function autoMarkObjective(examSubjectId: number) {
  return request.post<void>(`/marking/auto-mark/${examSubjectId}`)
}

// ===================== 阅卷码访问 API =====================

export interface MarkingSessionVO {
  sessionToken: string
  taskId: Id
  examName: string
  subjectName: string
  questionNo: string
  fullScore: number
  markingRole: number
  markingRoleName: string
  totalCount: number
  completedCount: number
  pendingCount: number
}

export interface MarkingItemVO {
  recordId: number
  answerSheetId: number
  questionId: number
  questionImage: string
  fullScore: number
  questionNo: string
  currentIndex: number
  totalCount: number
  annotations?: string
}

export interface ScoreSubmitDTO {
  recordId: number
  score: number
  comment?: string
  annotations?: string
}

// 阅卷码登录（免登录）
export function accessCodeLogin(accessCode: string) {
  return request.post<MarkingSessionVO>('/marking/access/login', { accessCode })
}

// 获取任务信息
export function getAccessTaskInfo(sessionToken: string) {
  return request.get<MarkingSessionVO>('/marking/access/task', {
    headers: { 'X-Session-Token': sessionToken }
  })
}

// 获取下一份待阅记录
export function getNextAccessItem(sessionToken: string) {
  return request.get<MarkingItemVO>('/marking/access/next', {
    headers: { 'X-Session-Token': sessionToken }
  })
}

// 提交评分
export function submitAccessScore(sessionToken: string, data: ScoreSubmitDTO) {
  return request.post<boolean>('/marking/access/submit', data, {
    headers: { 'X-Session-Token': sessionToken }
  })
}

// 跳过当前记录
export function skipAccessRecord(sessionToken: string, recordId: number) {
  return request.post<boolean>('/marking/access/skip', { recordId }, {
    headers: { 'X-Session-Token': sessionToken }
  })
}

// 生成阅卷码（需要管理员权限）
export function generateAccessCode(taskId: Id) {
  return request.post<MarkingTaskVO>(`/marking/access/generate/${taskId}`)
}

// 刷新阅卷码有效期
export function refreshAccessCode(taskId: Id, hours: number = 24) {
  return request.post<void>(`/marking/access/refresh/${taskId}?hours=${hours}`)
}

// 获取任务详情（含阅卷码）
export function getTaskWithAccessCode(taskId: Id) {
  return request.get<MarkingTaskVO>(`/marking/access/task/${taskId}`)
}

File: edumark-flask-server/app/api/admin/marking.py
from app.api.admin.common import *


@admin_bp.get("/marking/task/page")
@login_required
def marking_task_page():
    items, total, page_num, page_size = store.page("marking_tasks", request.args.to_dict(), [])
    return page_result([_marking_task_payload(item) for item in items], total, page_num, page_size)


@admin_bp.get("/marking/task/<int:item_id>")
@login_required
def marking_task_detail(item_id: int):
    item = store.find_by_id("marking_tasks", item_id)
    if not item:
        return ok({})
    return ok(_marking_task_payload(item, include_assigns=True))


@admin_bp.post("/marking/task/generate/<int:exam_subject_id>")
@login_required
def marking_task_generate(exam_subject_id: int):
    item = store.create("marking_tasks", {"examSubjectId": exam_subject_id, "status": 1, "statusName": "待分配", "taskName": f"阅卷任务-{exam_subject_id}", "questionId": 1000 + exam_subject_id, "questionNo": 1, "assignCount": 0, "completedCount": 0})
    return ok(item["id"])


@admin_bp.delete("/marking/task/<int:item_id>")
@login_required
def marking_task_delete(item_id: int):
    store.delete("marking_tasks", item_id)
    return ok()


@admin_bp.post("/marking/task/assign")
@login_required
def marking_task_assign():
    payload = _json()
    task = store.find_by_id("marking_tasks", payload.get("taskId"))
    if not task:
        return fail("阅卷任务不存在")
    assigns = payload.get("assigns") or []
    if assigns:
        first = assigns[0]
        teacher = store.find_by_id("teachers", first.get("teacherId")) or {}
        task["teacherId"] = first.get("teacherId")
        task["teacherName"] = teacher.get("teacherName", "")
        task["assignCount"] = first.get("assignCount", task.get("assignCount", 0))
    task["status"] = 1
    task["statusName"] = "待开始"
    return ok(message="阅卷任务分配完成")


@admin_bp.post("/marking/task/start/<int:item_id>")
@login_required
def marking_task_start(item_id: int):
    task = store.find_by_id("marking_tasks", item_id)
    if task:
        task["status"] = 2
        task["statusName"] = "进行中"
    return ok()


@admin_bp.post("/marking/task/complete/<int:item_id>")
@login_required
def marking_task_complete(item_id: int):
    task = store.find_by_id("marking_tasks", item_id)
    if task:
        task["status"] = 3
        task["statusName"] = "已完成"
    return ok()


@admin_bp.get("/marking/task/list/<int:exam_subject_id>")
@login_required
def marking_task_list(exam_subject_id: int):
    items = [item for item in store.list("marking_tasks") if item.get("examSubjectId") == exam_subject_id]
    return ok([_marking_task_payload(item) for item in items])


@admin_bp.get("/marking/my-assigns")
@login_required
def marking_my_assigns():
    return ok([_marking_assign_payload(task) for task in store.list("marking_tasks")])


@admin_bp.get("/marking/records")
@login_required
def marking_records():
    items, total, page_num, page_size = store.page("marking_records", request.args.to_dict(), [])
    return page_result([_marking_record_payload(item) for item in items], total, page_num, page_size)


@admin_bp.get("/marking/next-pending")
@login_required
def marking_next_pending():
    record = store.list("marking_records")[0] if store.list("marking_records") else None
    return ok(_marking_record_payload(record) if record else None)


@admin_bp.get("/marking/record/<int:record_id>")
@login_required
def marking_record_detail(record_id: int):
    record = store.find_by_id("marking_records", record_id) or {}
    return ok(_marking_record_payload(record) if record else {})


@admin_bp.post("/marking/submit")
@login_required
def marking_submit():
    payload = _json()
    record = store.find_by_id("marking_records", payload.get("recordId"))
    if record:
        record["score"] = payload.get("score", record.get("score"))
        record["comment"] = payload.get("comment", "")
        record["status"] = 1
        record["statusName"] = "已完成"
    return ok(message="阅卷提交成功")


@admin_bp.get("/marking/arbitrations")
@login_required
def marking_arbitrations():
    items, total, page_num, page_size = store.page("marking_arbitrations", request.args.to_dict(), [])
    return page_result([_marking_arbitration_payload(item) for item in items], total, page_num, page_size)


@admin_bp.get("/marking/next-arbitration")
@login_required
def marking_next_arbitration():
    arbitration = store.list("marking_arbitrations")[0] if store.list("marking_arbitrations") else None
    return ok(_marking_arbitration_payload(arbitration) if arbitration else None)


@admin_bp.get("/marking/arbitration/<int:arbitration_id>")
@login_required
def marking_arbitration_detail(arbitration_id: int):
    item = store.find_by_id("marking_arbitrations", arbitration_id)
    return ok(_marking_arbitration_payload(item) if item else {})


@admin_bp.post("/marking/arbitration/submit")
@login_required
def marking_arbitration_submit():
    return ok(message="仲裁提交成功")


@admin_bp.post("/marking/auto-mark/<int:exam_subject_id>")
@login_required
def marking_auto_mark(exam_subject_id: int):
    return ok(message="自动阅卷已触发")


@admin_bp.post("/marking/access/login")
def marking_access_login():
    task = store.find_by_id("marking_tasks", 1) or {}
    return ok(_marking_session_payload(task))


@admin_bp.get("/marking/access/task")
def marking_access_task():
    task = store.find_by_id("marking_tasks", 1) or {}
    return ok(_marking_session_payload(task))


@admin_bp.get("/marking/access/next")
def marking_access_next():
    record = store.list("marking_records")[0] if store.list("marking_records") else {}
    return ok(_marking_item_payload(record))


@admin_bp.post("/marking/access/submit")
def marking_access_submit():
    return ok(True)


@admin_bp.post("/marking/access/skip")
def marking_access_skip():
    return ok(True)


@admin_bp.post("/marking/access/generate/<int:task_id>")
@login_required
def marking_access_generate(task_id: int):
    task = store.find_by_id("marking_tasks", task_id)
    if not task:
        return fail("阅卷任务不存在")
    task["accessCode"] = f"AC{task_id:04d}"
    task["accessCodeExpireTime"] = store.now()
    return ok(_marking_task_payload(task))


@admin_bp.post("/marking/access/refresh/<int:task_id>")
@login_required
def marking_access_refresh(task_id: int):
    return ok()


@admin_bp.get("/marking/access/task/<int:task_id>")
@login_required
def marking_access_task_detail(task_id: int):
    task = store.find_by_id("marking_tasks", task_id)
    return ok(_marking_task_payload(task, include_assigns=True) if task else {})
